### PR TITLE
Implement Limit on Resource-Intensive Batch Requests in JSON RPC

### PR DESCRIPTION
**Pull request by Gabriel-Trintinalia**
## PR description

This Pull Request introduces a new limitation mechanism for batch requests in Besu's JSON RPC that are deemed resource-intensive. Non-resource-intensive methods can be called without restrictions, while resource-intensive requests are now subject to a limit, which is defined by Besu's `JsonRpcConfiguration`. Currently, the `eth_call` method is the only one marked as resource-intensive.

Changes

- Introduced a limit on the number of resource-intensive requests that can be executed in a batch of JSON RPC requests within Besu. The limit is configurable.
- Implemented an optional command-line interface parameter `rpc-http-max-resource-intensive-per-batch-size` allowing users to manually adjust the resource-intensive request limit as per their needs.
- Established a new error response mechanism. If the number of resource-intensive requests exceeds the set limit, all subsequent resource-intensive requests will receive a specific error.

For instance, given the limit of 1 for the resource-intensive eth_call method, the response to this batch of requests would be:

Request
```
[
  {"jsonrpc":"2.0", "method":"eth_call", "params":[{/*parameters for eth_call*/}], "id":1},
  {"jsonrpc":"2.0", "method":"eth_call", "params":[{/*parameters for eth_call*/}], "id":2},
  {"jsonrpc":"2.0", "method":"eth_getTransactionCount", "params":[{/*parameters for eth_getTransactionCount*/}], "id":3},
  {"jsonrpc":"2.0", "method":"eth_getBlockByNumber", "params":[{/*parameters for eth_getBlockByNumber*/}], "id":4}
]
```
Response:
```
[
  {
    "jsonrpc": "2.0",
    "id": 1,
    "result": { /* result of first eth_call */ }
  },
  {
    "jsonrpc": "2.0",
    "id": 2,
    "error": {
      "code": -32005,
      "message": "Resource-Intensive Request Limit Exceeded",
    }
  },
  {
    "jsonrpc": "2.0",
    "id": 3,
    "result": { /* result of eth_getTransactionCount */ }
  },
  {
    "jsonrpc": "2.0",
    "id": 4,
    "result": { /* result of eth_getBlockByNumber */ }
  }
]
```

### CODE REVIEW

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/handlers/batch/JsonRpcBatchExecutor.java [hunk 1]
```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright ConsenSys AG.
```
Comment (macfarla) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/handlers/batch/JsonRpcBatchExecutor.java:2]:
```suggestion
 * Copyright contributors to Hyperledger Besu.
```

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/response/JsonRpcError.java [hunk 2]
```diff
@@ -60,6 +60,9 @@ public enum JsonRpcError {
   EXCEEDS_BLOCK_GAS_LIMIT(-32005, "Transaction gas limit exceeds block gas limit"),
   EXCEEDS_RPC_MAX_BLOCK_RANGE(-32005, "Requested range exceeds maximum RPC range limit"),
   EXCEEDS_RPC_MAX_BATCH_SIZE(-32005, "Number of requests exceeds max batch size"),
+  EXCEEDS_RPC_MAX_RESOURCE_INTENSIVE_BATCH_SIZE(
+      -32005, "Resource-Intensive Request Limit Exceeded"),
```
Comment (macfarla) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/response/JsonRpcError.java:64]:
```suggestion
      -32005, "Number of resource-intensive requests exceeds batch limit"),
```

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/MaxResourceIntensivePerBatchJsonRpcHttpServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright ConsenSys AG.
```
Comment (macfarla) [inline ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/MaxResourceIntensivePerBatchJsonRpcHttpServiceTest.java:2]:
```suggestion
 * Copyright contributors to Hyperledger Besu.
```

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/MaxResourceIntensivePerBatchJsonRpcHttpServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    }
+  }
+
+  // Create a batch request containing two types of requests: one resource-intensive and one
+  // non-resource-intensive.
+  // The batch request is structured as a JSON array containing five requests in total.
+  private RequestBody createRequestBody() {
+    return RequestBody.create(
+        "["
+            // 1 - Non-resource-intensive request
```
Comment (macfarla) [inline ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/MaxResourceIntensivePerBatchJsonRpcHttpServiceTest.java:108]: I don't think we need these comments 1-5 - the method name is a good explanation

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/MaxResourceIntensivePerBatchJsonRpcHttpServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+
+  private void initMaxResourceIntensivePerBatchSize(final int rpcResourceIntensiveMaxBatchSize)
+      throws Exception {
+    maxResourceIntensivePerBatchSize = rpcResourceIntensiveMaxBatchSize;
+    resourceIntensiveMethods = List.of("resourceIntensiveMethod");
+    initServerAndClient();
+  }
+
+  @Test
+  public void shouldOnlyExecuteOneResourceRequestOPerBatchWhenConfigIsEnabled() throws Exception {
```
Comment (macfarla) [inline ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/MaxResourceIntensivePerBatchJsonRpcHttpServiceTest.java:41]:
```suggestion
  public void shouldOnlyExecuteOneResourceIntensiveRequestPerBatch_whenLimitIsEnabled() throws Exception {
```

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/MaxResourceIntensivePerBatchJsonRpcHttpServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+          .isEqualTo(JsonRpcError.EXCEEDS_RPC_MAX_RESOURCE_INTENSIVE_BATCH_SIZE.getCode());
+
+      // Assert that the third non-resource-intensive request executed successfully (it should not
+      // contain an error).
+      assertThat(json.getJsonObject(4).getString("error")).isNull();
+    }
+  }
+
+  @Test
+  public void shouldOnlyExecuteOneResourceRequestOPerBatchWhenConfigIsDisabled() throws Exception {
```
Comment (macfarla) [inline ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/MaxResourceIntensivePerBatchJsonRpcHttpServiceTest.java:77]:
```suggestion
  public void shouldExecuteAllResourceIntensiveRequestsInBatch_whenLimitIsDisabled() throws Exception {
```

**Review by macfarla — APPROVED** [review]
This is a pretty neat solution!

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java [hunk 7]
```diff
@@ -767,6 +767,14 @@ static class JsonRPCHttpOptionGroup {
         description =
             "Specifies the maximum number of requests in a single RPC batch request via RPC. -1 specifies no limit  (default: ${DEFAULT-VALUE})")
     private final Integer rpcHttpMaxBatchSize = DEFAULT_HTTP_MAX_BATCH_SIZE;
+
+    @CommandLine.Option(
+        names = {"--rpc-http-max-resource-intensive-per-batch-size"},
+        paramLabel = MANDATORY_INTEGER_FORMAT_HELP,
+        description =
+            "Specifies the maximum number of resource-intensive requests in a single RPC batch request via RPC. -1 specifies no limit  (default: ${DEFAULT-VALUE})")
```
Comment (fab-10) [inline besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java:775]: For the UX will be nice to have a way to know which are the resource intensive methods

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/handlers/batch/JsonRpcBatchExecutor.java [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+   * @return A JsonRpcResponse corresponding to the result of the request.
+   */
+  private JsonRpcResponse handleResourceIntensiveRequest(final JsonObject rpcRequest) {
+    if (canProcessResourceIntensiveRequest(rpcRequest)) {
+      resourceIntensiveRequestsCounter++;
+      return executeRpcRequest(rpcRequest);
+    } else {
+      // If the request is resource-intensive and the limit for such requests has been
+      // exceeded, retrieve the request ID and add an error response
+      final Integer requestId = rpcRequest.getInteger("id", null);
```
Comment (fab-10) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/handlers/batch/JsonRpcBatchExecutor.java:125]: will be nice to have a metrics for this kind of error, if not already present

**Review by fab-10 — COMMENTED** [review]
LGTM, just a couple of wishes, and waiting for the answer to the Simon questions

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/handlers/batch/JsonRpcBatchExecutor.java (hunk 8 above)
Comment (macfarla) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/handlers/batch/JsonRpcBatchExecutor.java:125]: agree metrics would be nice - could be a separate PR though

**Pull request closed without merging**